### DISCUSSION TITLE
Why is everything OneOrMany<T>, even things that would logically only have 1 value?

### DISCUSSION THREAD

**Discussion by hansmbakker** [RehanSaeed/Schema.NET#205] [Q&A] [answered]
### Question

In `Recipe` I see that [`Cooktime` has a type of `OneOrMany<TimeSpan?>`](https://github.com/RehanSaeed/Schema.NET/blob/master/Source/Schema.NET/core/Recipe.cs#L20). Logically, a recipe would not have more than one cooking times.

It seems all properties are marked as `OneOrMany<T>`; is that done intentionally? Why?

### Schema Objects

Which [schema.org](https://schema.org) object is this about if any?

- https://schema.org/Recipe

**Comment by Turnerj** [comment, marked as answer]
Each of the schema types are generated by a tool, which takes a JSON file from schema.org. The JSON file doesn't specify whether some properties are logically only ever a single value or not.

You will find a lot of properties like this though there are also many properties where multiples could exist (eg. author). Without some kind of mapping to determine what makes logical sense or not, defaulting to the idea that anything could be multiple values makes life a lot easier.

When dealing with `OneOrMany` though, you don't have to worry too much about it being a single value or many values. It has [implicit conversions from a singular `T` to a `OneOrMany<T>`](https://github.com/RehanSaeed/Schema.NET/blob/095918195d9a7fc6d15bce9251e556ecbe435636/Source/Schema.NET/OneOrMany%7BT%7D.cs#L146). It has these implicit conversions [the opposite way](https://github.com/RehanSaeed/Schema.NET/blob/095918195d9a7fc6d15bce9251e556ecbe435636/Source/Schema.NET/OneOrMany%7BT%7D.cs#L175) too (`OneOrMany<T>` to `T`).

Hopefully that answers your question!

**Comment by hansmbakker** [comment]
Ok, that is clear. Out of interest, though: do you know whether there are many apps that use these schema.org models directly as their view models?

Or is it more common to have converters that convert between the schema.org models and view models (I suspect this, but not sure)?

**Comment by Turnerj** [comment]
I don't have any firm numbers either way. I'd suspect for people who are generating JSON+LD for web pages, they might use the Schema object directly as it saves them from unnecessary conversion and serialization logic - this is what I do for my sites.

Schema.NET is smart enough to realise when you have one value in a `OneOrMany`, it will serialize the one value rather than an array with one value.

The only other example that comes to mind is another thing I'm working on, I have a product where the API returns Schema.org types where the API model itself is the Schema.NET type. I do this as I would have needed to re-implement every possible type in Schema.NET to have a stable view model. Instead, I decided whatever the latest official version of Schema.org is will be what I use.

**Comment by hansmbakker** [comment]
Thank you!
